persistence/pg.go

```diff
@@ -249,53 +263,87 @@ func (p *PostgresPersister) getHtlcHash(ctx context.Context, tx *pg.Tx,
 	return htlc.Hash, nil
 }
 
-func (p *PostgresPersister) MarkHtlcSettled(ctx context.Context,
-	key types.HtlcKey) (*lntypes.Hash, error) {
+func (p *PostgresPersister) MarkHtlcFinal(ctx context.Context,
+	key types.HtlcKey, settled bool) (*lntypes.Hash, error) {
+
+	var invoiceHash *lntypes.Hash
 
-	var settledHash *lntypes.Hash
+	var htlcFinalStatus types.HtlcStatus
+	if settled {
+		htlcFinalStatus = types.HtlcStatusSettled
+	} else {
+		htlcFinalStatus = types.HtlcStatusFailed
+	}
 
 	err := p.conn.RunInTransaction(ctx, func(tx *pg.Tx) error {
-		// Select invoice FOR UPDATE to prevent incorrect counts when multiple
-		// htlcs for the same invoice are marked as settled concurrently.
+		// Select invoice FOR UPDATE to prevent concurrent update on the
+		// invoice/htlc.
 		invoice, err := p.selectInvoiceForUpdate(ctx, tx, key)
 		if err != nil {
 			return err
 		}
 
-		now := time.Now().UTC()
-
-		// Settle the HTLC in the database.
-		if err := p.settleHTLC(ctx, tx, key, now); err != nil {
-			return err
+		// If invoice has InvoiceStatusSettled status, it means that all
+		// HTLC were settled earlier and MarkHtlcFinal should not have
+		// been called again for this invoice.
+		if invoice.Status == types.InvoiceStatusSettled {
+			return ErrHtlcReceivedButInvoiceAlreadySettled
 		}
 
-		// The HTLC is settled, now we check if all other HTLCs
-		// are settled as well.
-		allHtlcsSettled, err := p.allHtlcsAreSettled(ctx,
-			tx, invoice.Hash)
+		now := time.Now().UTC()
+
+		err = p.changeHTLCStatus(ctx, tx, key, htlcFinalStatus, now)
```

Comment:
nit: change sounds like you can change it to anything. Maybe `finalizeHTLC(settled)`?